### PR TITLE
V3.1

### PR DESCRIPTION
**Pull request by Aridjar**
one error happen when we set "call" to false (in purpose to test the application). Otherwise everything's fine.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -261,6 +261,37 @@ testEmail('Hello World!');
 npm test
 ```
 
+## Use the version 3.1 of the send api
```
Comment (arnaudbreton) [inline README.md]: `send api` => `Send API`

File: README.md [hunk 2]
```diff
@@ -261,6 +261,37 @@ testEmail('Hello World!');
...
+
+``` javascript
+
+// The third argument (the object) is not mandatory, as each of its 4 keys.
+const mailjet = require ('apiv3')
+    .connect(process.env.MJ_APIKEY_PUBLIC, process.env.MJ_APIKEY_PRIVATE, {
+        'url': 'api.mailjet.com', // default is the API url
+        'version': 'v3', // default is '/v3'
+        'secured': "https", // default is 'https'
+        'call': true // used for tests. default is true
```
Comment (arnaudbreton) [inline README.md]: `perform_api_call` to align with Ruby?

File: mailjet-client.js [hunk 3]
```diff
@@ -272,11 +292,14 @@ function MailjetResource (method, func, context) {
...
-    })())
+    })(), that.opts)
+    
+    secured = ('secured' in that.opts && that.opts['secured'] === "http" ? "http" : self.config.secured)
+    call = ('call' in that.opts && that.opts['call'] === false ? false : self.config.call)
 
     that.callUrl = that.base
     self.lastAdded = RESOURCE
-    return self.httpRequest(method, 'https://' + _path.join(self.config.url, path), params, callback)
+    return self.httpRequest(method, secured + '://' + path, params, callback, call)
```
Comment (arnaudbreton) [inline mailjet-client.js]: `secured` should be a boolean, not a string.

File: mailjet-client.js [hunk 4]
```diff
@@ -359,8 +382,8 @@ MailjetResource.prototype.request = function (params, callback) {
  *
  * @returns a function that make an httpRequest for each call
  */
-MailjetClient.prototype.post = function (func) {
-  return new MailjetResource('post', func, this)
+MailjetClient.prototype.post = function (func, opts) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Rename `opts` to `options`

File: mailjet-client.js [hunk 5]
```diff
@@ -171,11 +190,11 @@ MailjetClient.prototype.httpRequest = function (method, url, data, callback) {
     console.log('Final url: ' + url)
     console.log('body: ' + payload)
   }
-
-  if (this.testMode) {
+  
+  if (call === false || this.testMode) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: `if (call === false` => `if (!call`

File: mailjet-client.js [hunk 6]
```diff
@@ -106,9 +106,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
   this.apiKey = apiKey
   this.apiSecret = apiSecret
   this.options = options || {}
+  if (options) {
+    this.config = this.setConfig(options);
```
Comment (arnaudbreton) [inline mailjet-client.js]: Already done at line https://github.com/mailjet/mailjet-apiv3-nodejs/pull/21/files#diff-6d3c1ffe8a9e86b3bcaae3a208bd967dR63?

File: mailjet-client.js [hunk 7]
```diff
@@ -120,19 +135,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
...
-MailjetClient.prototype.path = function (resource, sub, params) {
+MailjetClient.prototype.path = function (resource, sub, params, opts) {
   if (DEBUG_MODE) {
     console.log('resource =', resource)
     console.log('subPath =', sub)
     console.log('filters =', params)
   }
-  var base = _path.join(this.config.version, sub)
+  
+  is_object = typeof opts === 'object' ? true : false
```
Comment (arnaudbreton) [inline mailjet-client.js]: Not required in JS I think

File: mailjet-client.js [hunk 8]
```diff
@@ -224,9 +243,10 @@ MailjetClient.prototype.httpRequest = function (method, url, data, callback) {
  * @func {String} resource/path to be sent
  * @context {MailjetClient[instance]} parent client
  */
-function MailjetResource (method, func, context) {
+function MailjetResource (method, func, opts = {}, context) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Signature not backward compatible. Impact on public interface?

File: mailjet-client.js (hunk 5 above)
Comment (lohek) [inline mailjet-client.js]:
```js
call = null
console.log(!call) // true

call = false
console.log(!call) // true
```

So maybe why `===`

Comment (Aridjar) [inline mailjet-client.js]: exact

File: config.json [hunk 9]
```diff
@@ -1,5 +1,7 @@
 {
-	"url": "api.mailjet.com",
-	"version": "/v3/",
-	"output": "json"
+  "url": "api.mailjet.com",
+  "version": "/v3/",
```
Comment (Aridjar) [inline config.json]: should be v3

File: README.md (hunk 1 above)
Comment (Aridjar) [inline README.md]: done

File: mailjet-client.js (hunk 6 above)
Comment (Aridjar) [inline mailjet-client.js]: Different call, the connect is here to overwrite the initial declaration. If we only want to modify during 1 function, we will use the option part of the different actions (post, put, get...)

File: README.md (hunk 2 above)
Comment (Aridjar) [inline README.md]: done. To be test

File: mailjet-client.js (hunk 8 above)
Comment (Aridjar) [inline mailjet-client.js]: This function is called by the wrapper. The user should never call it, and has to read the code to know it exist.

File: mailjet-client.js (hunk 4 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: config.json (hunk 9 above)
Comment (Aridjar) [inline config.json]: done

File: mailjet-client.js [hunk 10]
```diff
@@ -272,11 +292,23 @@ function MailjetResource (method, func, context) {
       } else {
         return {}
       }
-    })())
+    })(), that.opts)
+    
+    if (that.opts && that.opts.hasOwnProperty("secured")) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Simplify to `that.opts.secured`

File: mailjet-client.js [hunk 11]
```diff
@@ -272,11 +292,23 @@ function MailjetResource (method, func, context) {
...
+      secured = that.opts['secured'];
+    } else {
+      secured = self.config.secured;
+    }
+    
+    if (that.opts && that.opts.hasOwnProperty("perfom_api_call")) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Simplify to `that.opts.perform_api_call`

File: mailjet-client.js [hunk 12]
```diff
@@ -120,19 +135,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
  * @params {Object literal} {name: value}
  *
  */
-MailjetClient.prototype.path = function (resource, sub, params) {
+MailjetClient.prototype.path = function (resource, sub, params, opts) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Align opts to `options`

File: mailjet-client.js [hunk 13]
```diff
@@ -106,9 +106,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
...
+  }
   return this
 }
 
+MailjetClient.prototype.setConfig = function (options) {
+  config = require('./config')
+  if (typeof options === 'object' && options != null && options.length != 0) {
+    if ('url' in options) config['url'] = options['url']
+    if ('version' in options) config['version'] = options['version']
+    if ('secured' in options && options['secured'] === false) config['secured'] = false
```
Comment (arnaudbreton) [inline mailjet-client.js]: Set it anyway if set in the options dict

File: mailjet-client.js [hunk 14]
```diff
@@ -106,9 +106,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
...
+    if ('perform_api_call' in options && options['perform_api_call'] === false) config['perform_api_call'] = false
```
Comment (arnaudbreton) [inline mailjet-client.js]: Set it anyway if set in the options dict

File: mailjet-client.js [hunk 15]
```diff
@@ -106,9 +106,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
...
+    this.config = this.setConfig(options);
+  }
   return this
 }
 
+MailjetClient.prototype.setConfig = function (options) {
+  config = require('./config')
+  if (typeof options === 'object' && options != null && options.length != 0) {
+    if ('url' in options) config['url'] = options['url']
+    if ('version' in options) config['version'] = options['version']
```
Comment (arnaudbreton) [inline mailjet-client.js]: Use dot notation: `options.version`

File: mailjet-client.js [hunk 16]
```diff
@@ -106,9 +106,24 @@ MailjetClient.prototype.connect = function (apiKey, apiSecret, options) {
...
   this.options = options || {}
+  if (options) {
+    this.config = this.setConfig(options);
+  }
   return this
 }
 
+MailjetClient.prototype.setConfig = function (options) {
+  config = require('./config')
+  if (typeof options === 'object' && options != null && options.length != 0) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Find a way to send back an error to the user

File: mailjet-client.js [hunk 17]
```diff
@@ -171,11 +190,11 @@ MailjetClient.prototype.httpRequest = function (method, url, data, callback) {
     console.log('Final url: ' + url)
     console.log('body: ' + payload)
   }
-
-  if (this.testMode) {
+  
+  if (perform_api_call === false || this.testMode) {
```
Comment (arnaudbreton) [inline mailjet-client.js]: Rename `testMode` to `perform_api_call`

File: mailjet-client.js (hunk 15 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 13 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 14 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 12 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 17 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 10 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 11 above)
Comment (Aridjar) [inline mailjet-client.js]: done

File: mailjet-client.js (hunk 16 above)
Comment (Aridjar) [inline mailjet-client.js]: Done. Using throw

**Pull request merged**